### PR TITLE
Refactor `std::tx` to use `gtf` opcode

### LINKED ISSUE #2146: Update the std library to use `gtf` in place of hardcoded offsets
blocked by: implement `gtf` opcode in compiler.

### PR DESCRIPTION
**Pull request by nfurfaro**
Lots of changes in this one, but it won't compile yet as the `gtf` opcode hasn't been fully implemented yet.
- I've tried to avoid renaming _for the most part_ to minimize breaking changes.
- Some of the fields we need to read are exposed under more than one hex immediate, for example:
`GTF_INPUT_COIN_TX_ID` & `GTF_INPUT_CONTRACT_TX_ID` both read the same field, and this is a bit of a pattern.
The approach I've taken is to expose a single wrapper function which matches on the type of Input/Output, and returns an `Option::Some` if possible, else an `Option::None`. 
- I haven't exposed the individual getters directly (i.e: this has a `tx_input_owner()` function, but not `tx_input_coin_owner()` or `tx_input_message_owner()` functions.) but it wouldn't be hard to do so.
- Some follow-up PRs should probably split out a few different modules from `std::tx`while adding the remaining functions, as this thing is going to get pretty big with the addition of all the `gtf` variants (~64 wrapper functions in total).

blocked by: #2179 
closes #2146

### CODE REVIEW

**Comment by simonr0204** [comment]
What's the `index` required everywhere as a parameter? For example `tx_type(index: u64)`

File: sway-lib-std/src/tx.sw [hunk 1]
```diff
@@ -10,41 +10,95 @@ use ::mem::read;
...
+const GTF_SCRIPT_GAS_LIMIT = 0x003;
+const GTF_SCRIPT_MATURITY = 0x004;
+const GTF_SCRIPT_SCRIPT_LENGTH = 0x005;
+const GTF_SCRIPT_SCRIPT_DATA_LENGTH = 0x006;
+const GTF_SCRIPT_INPUTS_COUNT = 0x007;
+const GTF_SCRIPT_OUTPUTS_COUNT = 0x008;
+const GTF_SCRIPT_WITNESSES_COUNT = 0x009;
+const GTF_SCRIPT_RECEIPTS_ROOT = 0x00A;
+const GTF_SCRIPT_SCRIPT = 0x00B;
+const GTF_SCRIPT_SCRIPT_DATA = 0x00C;
```
Comment (nfurfaro) [inline sway-lib-std/src/tx.sw:28]: Now that I added these, I don't think I can even use them.

**Comment by simonr0204** [comment]
> > What's the index required everywhere as a parameter? For example tx_type(index: u64)
> 
> The `gtf` opcode requires an Imput/Output index as a param for all operation.

But since it's the same transaction type (and gas price, limit, script bytecode, etc) for all inputs and outputs, shouldn't we just hard-code 0 as the index for those fields that apply to the whole transaction?

**Comment by nfurfaro** [comment]
>But since it's the same transaction type (and gas price, limit, script bytecode, etc) for all inputs and outputs, shouldn't we just hard-code 0 as the index for those fields that apply to the whole transaction?

Possibly? I'm not 100% sure if input//output ordering is fixed (or even matters), but I don't know how a user is expected to know the index of the Input/Output they're interested in otherwise.
Hardcoding the index to `0` would certainly make things much simpler.

**Comment by simonr0204** [comment]
> > But since it's the same transaction type (and gas price, limit, script bytecode, etc) for all inputs and outputs, shouldn't we just hard-code 0 as the index for those fields that apply to the whole transaction?
> 
> Possibly? I'm not 100% sure if Input/Output ordering is fixed (or even matters), but I don't know how a user is expected to know the index of the Input/Output they're interested in otherwise. Hardcoding the index to `0` would certainly make things much simpler.

Input/output ordering is one thing, but my comment is more that there are field that are common across the whole _transaction_. E.g. the gas price doesn't have anything to do with a given input or output. It's a property of the transaction.

File: sway-lib-std/src/tx.sw [hunk 2]
```diff
@@ -60,120 +114,116 @@ pub const OUTPUT_VARIABLE = 4u8;
 pub const OUTPUT_CONTRACT_CREATED = 5u8;
 
 /// Get the transaction type.
-pub fn tx_type() -> u8 {
-    asm(r1, r2: TX_TYPE_OFFSET) {
-        lw r1 r2 i0;
-        r1: u8
+pub fn tx_type(index: u64) -> u8 {
```
Comment (mohammadfawaz) [inline sway-lib-std/src/tx.sw:117]: `index` here is not really required right? From the spec, `index` is only useful and required for some of the tx fields like `GTF_SCRIPT_INPUT_AT_INDEX`

File: sway-lib-std/src/tx.sw (hunk 1 above)
Comment (mohammadfawaz) [reply sway-lib-std/src/tx.sw:28]: I assume you're blocked by https://github.com/FuelLabs/sway/issues/810 for this right?

File: sway-lib-std/src/tx.sw (hunk 2 above)
Comment (nfurfaro) [reply sway-lib-std/src/tx.sw:117]: yup, exactly.
>looks like I need to give this another pass to find all such fields and hardcode the index to 0 for them (an index is mandatory for gtf usage).

**Comment by mohammadfawaz** [comment]
@nfurfaro https://github.com/FuelLabs/sway/pull/2315 is now in and you should be able to use the `__gtf` intrinsic instead of `asm` blocks. Some documentation for the intrinsic can be found in the description but please let me know if you have any questions!

**Pull request closed without merging**